src/core/qgsvectorlayer.cpp

```diff
@@ -1531,6 +1531,13 @@ void QgsVectorLayer::setDataSource( const QString &dataSource, const QString &ba
     }
 
     setLegend( QgsMapLayerLegend::defaultVectorLegend( this ) );
+
+    if ( mDataProvider->capabilities() & QgsVectorDataProvider::CreateLabeling )
+    {
+      std::unique_ptr< QgsAbstractVectorLayerLabeling > defaultLabeling( mDataProvider->createLabeling() );
```

Comment:
what's the advantage in using a `unique_ptr` here instead of directly passing the pointer to `setLabeling`?

Reply:
I think the next block should be protected by a

   if ( defaultLabeling )

block. (In case the provider doesn't return anything)

Reply:
On my end, I've gone down the road of what was done for createRenderer. I defer to @nyalldawson 's mastery of C++ :)

Reply:
@nirvn yeah, me too, I'm just trying to understand and learn something.

Reply:
I generally do this because it makes ownership explicit upfront - if anyone refactors in future there's no chance of introducing a leak.

Reply:
I see your point, just be aware of the not trivial overhead, I guess in this case it does not matter.

Reply:
Good point. This method is rarely called though, so I think it's ok. (There's also a chance that the compiler would optimise this trivial case anyway). 

But there's still an issue here - we need to explicitly guard against a nullptr return by the provider's function (which already can happen in the afs provider implementation)

Reply:
No optimizations for clang or gcc: https://godbolt.org/z/xLllBV